### PR TITLE
Add new sniff for checking Control structure brackets are on a new line

### PR DESCRIPTION
**Pull request by photodude**
Add new sniff for checking Control structure brackets are on a new line
Add new test for the new sniff checking Control structure brackets are on a new line

### CODE REVIEW

**Comment by photodude** [comment]
@gsherwood What Code standard should I use to format things here? I seem to have chosen the wrong one for the formatting.

**Comment by jrfnl** [comment]
@photodude PHPCS has its own ruleset you can use: https://github.com/squizlabs/PHP_CodeSniffer/blob/master/phpcs.xml.dist :smile:

**Comment by photodude** [comment]
Thanks @jrfnl, hopefully I implemented the fixer right since I just ran it directly from the github files so I didn't have to mess with my PHPCS 2.9 install.

**Comment by jrfnl** [comment]
@photodude Want me to have a look/review ?

**Comment by photodude** [comment]
Do you mean to review the sniff? Yes, please.
There are probably better was to do this sniff. The sniff was originally based off the PEAR.Classes.ClassDeclaration Sniff but modified to work with control structure brackets tokens.

I think the code style is fixed, if I can get the one comment section to stop yelling at me that it wants more or less indenting, or things at the end.

**Comment by photodude** [comment]
code style should be all in the green now

**Comment by photodude** [comment]
@jrfnl @gsherwood This is now good to go for review consideration

File: src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+     * @var integer
+     */
+    public $indent = 4;
+
+    /**
+     * A list of tokenizers this sniff supports.
+     *
+     * @var array
+     */
+    public $supportedTokenizers = ['PHP'];
```
Comment (gsherwood) [inline src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php]: This property isn't needed when you are only checking PHP files. It's the default state, so can be removed.

File: src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        $tokens    = $phpcsFile->getTokens();
+        $errorData = [strtolower($tokens[$stackPtr]['content'])];
+
+        if (isset($tokens[$stackPtr]['scope_opener']) === false) {
+            if ($tokens[$stackPtr]['code'] !== T_WHILE) {
+                $error = 'Possible parse error: %s missing opening or closing brace';
+                $phpcsFile->addWarning($error, $stackPtr, 'MissingBrace', $errorData);
```
Comment (gsherwood) [inline src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php]: This sniff should not put a warning here. They might just be using inline control structures, which another sniff can test for and ban.

File: src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+            if ($spaces !== $expected) {
+                $error = 'Expected %s tabs before opening brace; %s found';
+                $data  = [
+                    $expected,
+                    $spaces,
+                ];
+                $fix   = $phpcsFile->addFixableError($error, $openBrace, 'SpaceBeforeBrace', $data);
+
+                if ($fix === true) {
+                          $indent = str_repeat("\t", $expected);
```
Comment (gsherwood) [inline src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php]: Line indented too far.

File: src/Standards/Generic/Tests/ControlStructures/ControlStructuresBracketsUnitTest.inc [hunk 4]
```diff
@@ -0,0 +1,170 @@
+<?php
+
+$a = true;
+
+// Warning
```
Comment (gsherwood) [inline src/Standards/Generic/Tests/ControlStructures/ControlStructuresBracketsUnitTest.inc]: Test not needed (as per previous comment).

File: src/Standards/Generic/Tests/ControlStructures/ControlStructuresBracketsUnitTest.inc [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+			}
+			else
+		{
+				return StringHelper::strcmp($b->$ordering, $a->$ordering);
+			}
+		}
+	);
+}
+
+// Anonymous classes test
```
Comment (gsherwood) [inline src/Standards/Generic/Tests/ControlStructures/ControlStructuresBracketsUnitTest.inc]: There is a lot of real code in this test. I prefer to boil it down to only what is needed and then replace with fake code.

File: src/Standards/Generic/Tests/ControlStructures/ControlStructuresBracketsUnitTest.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+     * Returns the lines where warnings should occur.
+     *
+     * The key of the array should represent the line number and the value
+     * should represent the number of warnings that should occur on that line.
+     *
+     * @return array<int, int>
+     */
+    public function getWarningList()
+    {
+        return [6 => 1];
```
Comment (gsherwood) [inline src/Standards/Generic/Tests/ControlStructures/ControlStructuresBracketsUnitTest.php]: Not required, as per previous comments.

File: src/Standards/Generic/Tests/ControlStructures/ControlStructuresBracketsUnitTest.inc [hunk 7]
```diff
@@ -0,0 +1,170 @@
+<?php
```
Comment (gsherwood) [inline src/Standards/Generic/Tests/ControlStructures/ControlStructuresBracketsUnitTest.inc]: The test file doesn't have any examples of comments being placed between the closing parenthesis and the opening brace of the control structures, so it's not obvious how those would be fixed. A few examples of using `// comment` and the phpcs annotations like `// phpcs:ignore -- for reasons` would be good, just to make sure the sniff copes with them on other people's code.

File: src/Standards/Generic/Tests/ControlStructures/ControlStructuresBracketsUnitTest.inc (hunk 4 above)
Comment (photodude) [reply src/Standards/Generic/Tests/ControlStructures/ControlStructuresBracketsUnitTest.inc]: Removed

File: src/Standards/Generic/Tests/ControlStructures/ControlStructuresBracketsUnitTest.inc (hunk 7 above)
Comment (photodude) [reply src/Standards/Generic/Tests/ControlStructures/ControlStructuresBracketsUnitTest.inc]: I'll see about adding those cases

File: src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php (hunk 3 above)
Comment (photodude) [reply src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php]: Fixed

File: src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php (hunk 2 above)
Comment (photodude) [reply src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php]: Removed

File: src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php (hunk 1 above)
Comment (photodude) [reply src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php]: Removed

**Comment by photodude** [comment]
Quick question, 
I'm getting a CS error on "Implicit true comparisons prohibited; use === TRUE instead"
the error occurs for this comparison
```php
if (array_key_exists('nested_parenthesis', $tokens[$stackPtr])) {
```

What is the correct solution for this?

**Comment by gsherwood** [comment]
@photodude 
```
if (array_key_exists('nested_parenthesis', $tokens[$stackPtr]) === true) {
```

**Comment by jrfnl** [comment]
@photodude 

```php
if (array_key_exists('nested_parenthesis', $tokens[$stackPtr]) === true) {
```

**Comment by photodude** [comment]
Thanks @gsherwood (and @jrfnl I saw your comment too for a moment)

I believe the last items here are

- [ ] convert the unit test into fake code
- [ ] add unit test cases using `// comment` and the phpcs annotations like `// phpcs:ignore -- for reasons`

File: src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+            for ($next = ($opener + 1); $next < $phpcsFile->numTokens; $next++) {
+                $code = $tokens[$next]['code'];
+
+                if ($code === T_WHITESPACE) {
+                    continue;
+                }
+
+                // Skip all empty tokens on the same line as the opener.
+                if ($tokens[$next]['line'] === $tokens[$opener]['line']
+                    && (isset(PHP_CodeSniffer_Tokens::$emptyTokens[$code]) === true
```
Comment (gsherwood) [inline src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php]: Replace `PHP_CodeSniffer_Tokens` with `Tokens`
```suggestion
                    && (isset(Tokens::$emptyTokens[$code]) === true
```

File: src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+
+                          $phpcsFile->fixer->endChangeset();
+                }
+
+                return;
+            }//end if
+        }//end if
+
+        if ($tokens[($openBrace + 1)]['content'] !== $phpcsFile->eolChar) {
+            $error = 'Opening %s brace must be on a line by itself.';
```
Comment (gsherwood) [inline src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php]: Sniff errors don't end with full stops
```suggestion
            $error = 'Opening %s brace must be on a line by itself';
```

File: src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+                $phpcsFile->fixer->addNewlineBefore($openBrace);
+                $phpcsFile->fixer->endChangeset();
+            }
+
+            return;
+        } else {
+            $phpcsFile->recordMetric($stackPtr, 'Control Structure opening brace placement', 'new line');
+
+            if ($braceLine > ($controlStructureLine + 1)) {
+                $error = 'Opening brace of a %s must be on the line following the %s declaration.; Found %s line(s).';
```
Comment (gsherwood) [inline src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php]: Sniff errors don't end with full stops
```suggestion
                $error = 'Opening brace of a %s must be on the line following the %s declaration; found %s line(s)';
```

File: src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+    {
+        $tokens      = $phpcsFile->getTokens();
+        $errorData   = [strtolower($tokens[$stackPtr]['content'])];
+        $openBrace   = $tokens[$stackPtr]['scope_opener'];
+        $lastContent = $phpcsFile->findPrevious(T_WHITESPACE, ($openBrace - 1), $stackPtr, true);
+        $controlStructureLine = $tokens[$lastContent]['line'];
+        $braceLine            = $tokens[$openBrace]['line'];
+
+        if ($braceLine === $controlStructureLine) {
+            $phpcsFile->recordMetric($stackPtr, 'Control Structure opening brace placement', 'same line');
```
Comment (gsherwood) [inline src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php]: Just a case change
```suggestion
            $phpcsFile->recordMetric($stackPtr, 'Control structure opening brace placement', 'same line');
```

File: src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+                    $phpcsFile->fixer->replaceToken(($openBrace - 1), '');
+                }
+
+                $phpcsFile->fixer->addNewlineBefore($openBrace);
+                $phpcsFile->fixer->endChangeset();
+            }
+
+            return;
+        } else {
+            $phpcsFile->recordMetric($stackPtr, 'Control Structure opening brace placement', 'new line');
```
Comment (gsherwood) [inline src/Standards/Generic/Sniffs/ControlStructures/ControlStructuresBracketsNewLineSniff.php]: Just a case change
```suggestion
            $phpcsFile->recordMetric($stackPtr, 'Control structure opening brace placement', 'new line');
```

**Comment by gsherwood** [comment]
I've done another review of this but it's not ready to merge yet. It still needs those last two items completed (remove real code from tests + add tests for comment syntax) plus:
- I think this would be worth renaming to `BraceNewLine` to match other sniffs
- Please replace tabs with spaces in the unit tests files unless the tab is testing something

I've also put a few minor changes into a review for you.

File: src/Standards/Generic/Tests/ControlStructures/ControlStructuresBracketsUnitTest.inc (hunk 5 above)
Comment (photodude) [reply src/Standards/Generic/Tests/ControlStructures/ControlStructuresBracketsUnitTest.inc]: @gsherwood I hope these changes are sufficient to obfuscate the "real code" and make it focused on just the structure being tested.

**Comment by gsherwood** [comment]
I've done another review of this and it's not ready to merge:
- The unit test files don't match the sniff name, so they aren't actually running. The files need to be renamed so they are tested.
- If the files are renamed, the unit tests will fail because the list of line numbers to error counts is wrong.
- The sniff enforces indentation using tabs, which is not something other sniffs do. If the sniff wants to check indentation of the opening brace, it needs to do so using spaces. If you use tabs in your coding standard, the DisallowSpaceIndent sniff will go ahead and fix that for you, understanding where the proper tab stops are.
- The unit test files is indented using tabs, so this needs to be replaced with spaces.
- There is no need for an indent property in this sniff if you are just trying to align the opening brace under the IF statement. PHPCS provides 'column' indexes in the tokens array so you know where to align things. This will make the sniff possible to use for any indent type and length.

I'm going to remove this from the 3.5.0 release and there looks to be too much more to do here to get it ready.

**Comment by pimjansen** [comment]
@photodude any way that this will be finished? If you need some assistance let me know

**Comment by photodude** [comment]
@pimjansen I would love to get this finished. Assistance would be greatly appreciated and would greatly speed along the process. (I, unfortunately, have reduced availability to work on side projects like this at the moment)

**Comment by pimjansen** [comment]
What is the easiest way to help here since this is open for way too long. What is still open and the fastest way to continue on this branch?

@photodude @gsherwood

**Comment by photodude** [comment]
> What is the easiest way to help here since this is open for way too long. What is still open and the fastest way to continue on this branch?

If you can work on any of the open items in the [to do checklist](https://github.com/squizlabs/PHP_CodeSniffer/pull/1847#issuecomment-509692380) that would help complete this sniff.

**Comment by photodude** [comment]
@pimjansen I don't think I'm going to get back to working on this, as I'm now working in a different field. If you would like to take over and finish this PR all the necessary information on what needs to be completed is here.

**Comment by photodude** [comment]
Closing as Joomla4 is moving to PSR12. There is work to make this happen and I'm currently not working in PHP and don't know when I would be able to work on this again. 
Sadly others Who indicated they were willing to finish this PR did not submit anything for finishing this PR.
Overall PR is now stale and should be closed.

**Pull request closed without merging**